btx/processing/pipca.py

```diff
@@ -0,0 +1,1011 @@
+            )
+
+            pc_dist = np.linalg.norm(pcs[:5], axis=0)
+            std = np.std(pc_dist)
+            mu = np.mean(pc_dist)
+
+            batch_outliers = np.where(np.abs(pc_dist - mu) > std)[0] + n - m
+
+            self.outliers = (
+                np.concatenate((self.outliers, batch_outliers), axis=0)
+                if len(self.outliers)
+                else batch_outliers
+            )
+
+    def compression_loss(self, X, U, normalized=False):
+        """
+        Calculate the compression loss between centered observation matrix X
+        and its rank-q reconstruction.
+
+        Parameters
+        ----------
+        X : ndarray, shape (d x n)
+            flattened, centered image data from n run events
+        U : ndarray, shape (d x q)
+            first q singular vectors of X, forming orthonormal basis
+            of q-dimensional subspace of R^d
+
+        Returns
+        -------
+        Ln : float
+            compression loss of X
+        """
+        _, n = X.shape
+
+        UX = U.T @ X
+        UUX = U @ UX
+
+        Ln = ((np.linalg.norm(X - UUX, "fro")) ** 2) / n
+
+        if normalized:
+            Ln /= np.linalg.norm(X, "fro") ** 2
+
+        return L
+
+    def display_image(self, idx, output_dir="", save_image=False):
+        """
+        Method to retrieve single image from run subject to model binning constraints.
+
+        Parameters
+        ----------
+        idx : int
+            Run index of image to be retrieved.
+        output_dir : str, optional
+            File path to output directory, by default ""
+        save_image : bool, optional
+            Whether to save image to file, by default False
+        """
+
+        U, S, mu, var = self.get_model()
+
+        if self.rank != 0:
+            return
+
+        bin_factor = 1
+        if self.downsample:
+            bin_factor = self.bin_factor
+
+        n, q, m, d = self.get_params()
+
+        a, b, c = self.psi.det.shape()
+        b = int(b / bin_factor)
+        c = int(c / bin_factor)
+
+        fig, ax = plt.subplots(1)
+
+        counter = self.psi.counter
+        self.psi.counter = idx
+        img = self.get_formatted_images(1, 0, d)
+        self.psi.counter = counter
+
+        img = img - mu
+        img = np.reshape(img, (a, b, c))
+
+        pixel_index_map = retrieve_pixel_index_map(self.psi.det.geometry(self.psi.run))
+        binned_pim = bin_pixel_index_map(pixel_index_map, bin_factor)
+
+        img = assemble_image_stack_batch(img, binned_pim)
+
+        #vmin = np.min(img.flatten())
+        vmax = np.max(img.flatten())
+        ax.imshow(
+            img,
+            norm=colors.SymLogNorm(linthresh=1.0, linscale=1.0, vmin=0, vmax=vmax),
+            interpolation=None
+        )
+
+        if save_image:
+            plt.savefig(output_dir)
+
+        plt.show()
+
+    def verify_model_accuracy(self):
+        """
+        Run benchmarking to verify model accuracy.
+        """
```

Comment:
Let's put that with the "benchmark" discussion - maybe for another PR?

Reply:
The "benchmarking" referenced in the docstring of this method is different from the benchmarking performed in the benchmark script. This method runs batch PCA and PiPCA and compares the statistical accuracy of the models. Do we want to remove this method as well? If so the `compression_loss`  and `report_interval_data` methods will no longer be called and should probably be cleaned up as well. Thoughts @fredericpoitevin ?

Reply:
Yes let's get it out since it's not at the core of PiPCA but a method to evaluate it - I think it belongs with the other "benchmarking" PR. However let's try to refine the naming to avoid confusion between the two types of benchmarks.